### PR TITLE
Improve Dictionary Typescript Types

### LINKED ISSUE #3798: Improve Typescript Typings for Dictionary
Dictionary is tricky to type, since it is possible to run into naming collisions between its methods (put, delete, etc) and the dictionary element keys.  For now, we have landed on a solution where the developer will simply make an intersection between their defined dictionary structure and the given dictionary methods.

For example:

```java
interface MyDictionary{
  a:string;
  b?:number;
}class Journal extends Realm.Object {
  tags?: Realm.Dictionary & MyDictionary;
...
```

It would be desirable to handle this through generics and also make sure that the naming collisions are impossible to override.

### PR DESCRIPTION
**Pull request by takameyer**
## What, How & Why?
    
Basically making dictionaries more dynamic.  There is also a bit of refactoring done for the RealmInsertionModel to make it more readable

This closes #3798

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -572,20 +570,36 @@ type ExtractPropertyNamesOfType<T, PropType> = {
...
-type RealmCollectionRemappedModelPart<T> = {
-    [K in keyof T]?: T[K] extends Realm.Collection<infer GT> ? Array<GT | RealmInsertionModel<GT>> : never
+type RealmListsRemappedModelPart<T> = {
+    [K in ExtractPropertyNamesOfType<T, Realm.List<any>>]?: T[K] extends Realm.List<infer GT> ? Array<GT | RealmInsertionModel<GT>> : never
 }
 
 /**
- * Joins T stripped of all keys which value extends Realm.Collection and all inherited from Realm.Object,
- * with only the keys which value extends Realm.List or Realm.Set, remapped as Arrays.
+ * Exchanges properties defined as Realm.List<Model> with an optional Array<Model | RealmInsertionModel<Model>>.
```
Comment (kraenhansen) [inline types/index.d.ts:580]: This doc string probably needs updating.

File: types/index.d.ts [hunk 2]
```diff
@@ -236,20 +236,16 @@ declare namespace Realm {
 
 
      interface Dictionary{
```
Comment (nirinchev) [inline types/index.d.ts:238]: Shouldn't the interface declaration be generic since in Core you can constrain the dictionary values to be of specific type?

**Review by kneth — APPROVED** [review]

File: types/index.d.ts (hunk 2 above)
Comment (takameyer) [reply types/index.d.ts:238]: Yes, if they can be constrained, then we should definitely reflect that.

Comment (takameyer) [reply types/index.d.ts]: @nirinchev Dictionary now allows the value to be constrained

**Pull request merged**